### ISSUE TITLE
EKS IRSA support for S3 spot data feed

### ISSUE THREAD

**Issue by lmouhib** [kubecost/cost-analyzer-helm-chart#2238]
Thanks for your interest in Kubecost!  To ensure you get the help you need, please follow these guidelines:

- If you have questions about product functionality/usage or need assistance with product installation/configuration, please email support@kubecost.com, where our support engineering team will be happy to help!
- If you have a reproducible bug or a feature request, you're in the right place.  Please include the information requested below.

---
# For reproducible bugs, please provide the following

**Describe the bug**  
I am deploying Kubecost with AMP, I have a service account annotated with an IAM Role that allow access to the S3 bucket that has spot data feed. When I check the spot data feed diagnostic I get the following error:
`
operation error S3: ListObjects, https response error StatusCode: 403, RequestID: , HostID: , api error InvalidAccessKeyId: The AWS Access Key Id you provided does not exist in our records.`

The s3 client should have picked up the IRSA credentials and used them to read from the S3 bucket. Unless the issue seem to be from the sigv4 proxy. However, I see in the helm chart value the need to supply the aws credential for access key and secret key.

**To Reproduce**  
3. See error

**Expected behavior**  

**Screenshots**  

**Collect logs (please complete the following information):**
- Run `helm ls` and paste the output here:
- If the pod is stuck in init, run `kubectl logs <kubecost-cost-analyzer pod name> -n kubecost -c cost-analyzer-init` and paste output here:
- If a page is broken or not loading, open your javascript developer console and paste the ouput here:
- Run `kubectl logs <kubecost-cost-analyzer pod name> -n kubecost -c cost-analyzer-server` and paste the output here:
- Run `kubectl logs <kubecost-cost-analyzer pod name> -n kubecost -c cost-model` and paste the output here:
- Run `kubectl logs <kubecost-cost-analyzer pod name> -n kubecost -c cost-analyzer-frontend` and paste the output here:

**Please share the support case, if any**  
This is most commonly a Zendesk ticket number or link.

**What impact will this have on your ability to get value out of Kubecost?**

---
# For feature requests, please provide the following:

**What problem are you trying to solve?**  

**Describe the solution you'd like**  

**Describe alternatives you've considered**  

**Please share the support case, if any**  
This is most commonly a Zendesk ticket number or link.

**How would users interact with this feature?**

**What impact will this have on your ability to get value out of Kubecost?**

**Comment by thomasvn** [comment]
@lmouhib Thanks for reaching out! Here is a doc on how to configure IRSA (IAM Roles for Service Accounts) with Kubecost:
https://docs.kubecost.com/install-and-configure/install/multi-cluster/aws/long-term-storage-aws#method-2-attach-iam-role-to-service-account-method

Notably, it is the kubecost-cost-analyzer pod that is attempting to `S3: ListObjects` and will need the [serviceaccount](https://github.com/kubecost/cost-analyzer-helm-chart/blob/v1.103/cost-analyzer/templates/cost-analyzer-service-account-template.yaml#L9-L12) created.

**Comment by lmouhib** [comment]
I verified the jwt token and IAM role where available as ENV variable. This looks like a transient issue. I rebuilt the cluster from scratch and it worked fine.

**Comment by thomasvn** [comment]
@lmouhib Great to hear things are working now!

Could you clarify how you resolved it? Was it via [`.Values.serviceaccount.annotations`](https://github.com/kubecost/cost-analyzer-helm-chart/blob/v1.103/cost-analyzer/templates/cost-analyzer-service-account-template.yaml#L9-L12) or through an ENV variable passed to kubecost-cost-model container?

**Comment by rcasaleiro** [comment]
I am having a similar issue, I am passing the irsa annotation to the service account and I can confirm the serviceaccount has the annotation but it seems the kubecost-cost-analyzer pod is not using it.  
No service account defined on the pods `serviceAccountName` field, so it might be that if the cost-model uses the pod attached service account.
The cost model reports:
```ETL: CloudUsage[X/s3://X]: Run[Kulro]: QueryCloudUsage error for [2023-05-10T00:00:00+0000, 2023-05-13T00:00:00+0000): Sevice Key Secret not configured for account X
```

**Comment by lmouhib** [comment]
@thomasvn I have used ekstctl to create annotate the service account, as described in this [blog](https://blog.kubecost.com/blog/aws-amp-kubecost-integration/) . I also made sure that role annotated has access to the S3 bucket.

**Comment by lmouhib** [comment]
@thomasvn this error still persist, I am using version 1.103.2 of kubecost

**Comment by thomasvn** [comment]
@lmouhib Oh, I'm sorry if I misinterpreted your last comment thinking this was resolved!

You mention it's a transient issue. Can you provide more detail on when IRSA seems to be working correctly. And when IRSA is not working correctly?

**Comment by lmouhib** [comment]
I cannot tell why IRSA sometime work and other time its not working. Here is a sample of different errors I see in the logs from different deployments, 1.102 and 1.103. In cloudtrail, I see the IRSA role has been assumed, and when I do a `describe` on the pod, I see the JWT token path and the role ARN, when I inspect the JWT token path I find the actual file with the token. 

```
2023-05-18T14:12:06.283614538Z ERR Failed to lookup reserved instance data: no reservation data available in Athena
2023-05-18T14:12:06.283654383Z DBG Running Query: SELECT
        line_item_usage_start_date,
        savings_plan_savings_plan_a_r_n,
        line_item_resource_id,
        savings_plan_savings_plan_rate
    FROM cur_data as cost_data
    WHERE line_item_usage_start_date BETWEEN date '2023-05-17' AND date '2023-05-18'
    AND line_item_line_item_type = 'SavingsPlanCoveredUsage' ORDER BY
    line_item_usage_start_date DESC
2023-05-18T14:12:06.300874241Z ERR Failed to lookup savings plan data: Error fetching Savings Plan Data: QueryAthenaPaginated: start query error: operation error Athena: StartQueryExecution, https response error StatusCode: 400, RequestID: 3b95ea47-f213-4072-bf87-539c737934a7, api error UnrecognizedClientException: The security token included in the request is invalid.
```

```
cost-model 2023-05-18T13:56:08.189195605Z ERR Failed to lookup reserved instance data: no reservation data available in Athena
cost-model 2023-05-18T13:56:08.205261239Z ERR Failed to lookup savings plan data: Error fetching Savings Plan Data: QueryAthenaPaginated: start query error: operation error Athena: StartQueryExecution, https response error StatusCode: 400, RequestID: d43128b7-2ba9-4e16-8f31-e55684f3f125, api error UnrecognizedClientException: The security token included in the request is invalid.
```

```
cost-model 2023-05-18T13:56:13.838104786Z ERR ETL: CloudUsage[073430591850/s3://BUCKET-NAME]: Build[uDvDr]: QueryCloudUsage error: Sevice Key Secret not configured for account ACCOUNT-ID

```

**Comment by lmouhib** [comment]
When I look in the S3 bucket server access log, the request coming to the bucket with spot feed data are all `403` and the reason is `InvalidAccessKeyId`

**Comment by lmouhib** [comment]
@thomasvn I have run this multiple time and the error is with the `cost-model` container, I am not sure how the access key is handled but thats what it breaking the integration. The API calls are not even logged in Cloudtrail because they fail before being submitted to AWS endpoints.

Seems like there is a need to debug this with breakpoints to understand what is happening. My guess would be after, the `AssumeRoleWithWebIdentity` happens, the access key that is returned is not passed correctly to initialize the different AWS clients.

**Comment by lmouhib** [comment]
I tried to run the above on `open cost` with spot feed and it works without issues. I am not sure what is breaking with the kubecost build

**Comment by thomasvn** [comment]
@lmouhib Thanks for all the investigation you've done so far!

I took a closer look at this log:

```
cost-model 2023-05-18T13:56:13.838104786Z ERR ETL: CloudUsage[073430591850/s3://BUCKET-NAME]: Build[uDvDr]: QueryCloudUsage error: Sevice Key Secret not configured for account ACCOUNT-ID
```

This error error can only occur when the following condition is _not_ met:

```
aws.cloudConfig.ServiceKeyName == "" && aws.cloudConfig.ServiceKeySecret == ""
```

Could you run `helm get values kubecost` to make sure the `kubecostProductConfigs.awsServiceKeyName` and `kubecostProductConfigs.awsServiceKeyPassword` are not set? Could you also ensure that you don't have [`secret/cloud-service-key`](https://github.com/kubecost/cost-analyzer-helm-chart/blob/v1.103/cost-analyzer/templates/aws-service-key-secret.yaml#L13-L17) created in your namespace?

**Comment by lmouhib** [comment]
Thanks for getting back, I checked and the secret is not created. Here are the values from `helm`. I do not see any of the values you mentioned.

```
USER-SUPPLIED VALUES:
    affinity: {}
    global:
      amp:
        enabled: true
        prometheusServerEndpoint: http://localhost:8005/workspaces/ws-redacted
        remoteWriteService: https://aps-workspaces.region.amazonaws.com/workspaces/ws-redacted/api/v1/remote_write
        sigv4:
          region: region
      grafana:
        enabled: false
        proxy: false
    ingress:
      annotations:
        kubernetes.io/ingress.class: nginx
      enabled: false
      hosts:
      - cost-analyzer.local
      paths:
      - /
      tls: []
    kubecostFrontend:
      image: public.ecr.aws/kubecost/frontend
      imagePullPolicy: Always
      resources:
        requests:
          cpu: 10m
          memory: 55Mi
    kubecostModel:
      etl: true
      etlStoreDurationDays: 120
      extraEnv:
      - name: LOG_LEVEL
        value: trace
      image: public.ecr.aws/kubecost/cost-model
      imagePullPolicy: Always
      maxQueryConcurrency: 5
      resources:
        requests:
          cpu: 200m
          memory: 55Mi
      utcOffset: "+00:00"
      warmCache: true
      warmSavingsCache: true
    kubecostProductConfigs:
      athenaBucketName: s3://location
      athenaDatabase: athenacurcfn_c_u_r
      athenaProjectID: "ACCOUNT_ID"
      athenaRegion: REGION
      athenaTable: cur_data
      athenaWorkgroup: WORKGROUP
      awsSpotDataBucket: BUCKET_NAME
      awsSpotDataPrefix: spot-feed
      awsSpotDataRegion: REGION
      clusterName: CLUSTER_NAME
      projectID: "ACCOUNT_ID"
    networkPolicy:
      enabled: false
    nodeSelector: {}
    persistentVolume:
      dbSize: 32.0Gi
      enabled: true
      size: 32Gi
    podSecurityPolicy:
      enabled: false
    pricingCsv:
      enabled: false
      location:
        URI: s3://kc-csv-test/pricing_schema.csv
        csvAccessCredentials: pricing-schema-access-secret
        provider: AWS
        region: us-east-1
    priority:
      enabled: false
    prometheus:
      configmapReload:
        prometheus:
          enabled: false
          extraArgs: {}
          extraConfigmapMounts: []
          extraVolumeDirs: []
          image:
            pullPolicy: IfNotPresent
            repository: public.ecr.aws/bitnami/configmap-reload
            tag: 0.7.1
          name: configmap-reload
          resources: {}
      kube-state-metrics:
        disabled: false
      nodeExporter:
        enabled: false
      server:
        extraArgs:
          query.max-concurrency: 1
          query.max-samples: 100000000
        global:
          evaluation_interval: 1m
          external_labels:
            cluster_id: cluster-one
          scrape_interval: 1m
          scrape_timeout: 10s
        image:
          repository: public.ecr.aws/kubecost/prometheus
          tag: v2.35.0
        persistentVolume:
          enabled: true
          size: 32Gi
        resources: {}
        tolerations: []
    reporting:
      productAnalytics: false
    service:
      annotations: {}
      labels: {}
      port: 9090
      targetPort: 9090
      type: ClusterIP
    serviceAccount:
      annotations: {}
      create: true
    sigV4Proxy:
      host: aps-workspaces.REGION.amazonaws.com
      region: REGION
    tolerations: []
    
```

**Comment by lmouhib** [comment]
This is a screenshot of the pricing configmap 
![Screenshot_2023-05-19_at_21_31_57](https://github.com/kubecost/cost-analyzer-helm-chart/assets/4688305/9b89491b-66ef-4b21-bcb7-f509ec22dc1e)

**Comment by thomasvn** [comment]
Thanks for sharing those configs. I think we'll need to gather more information about your environment to investigate. Could you send the following to support@kubecost.com?

* [Bug report](https://docs.kubecost.com/troubleshooting/bug-report)
* `eksctl` commands you ran to create the IAM service account
* `kubectl describe serviceaccount -n kubecost`
* Link to this Github Issue

**Comment by patrickli-foxtel** [comment]
So I was scratching my head on this too. A related issue is that if the AWS CUR bucket is in a different account and you have to use `masterPayerARN`, setting them in `kubecostProductConfigs` never work, I can only use cloud integration setup. It turns out that kubecost will generate a `aws.json` file in `/var/configs` that fills some garbage default values if they are not set in helm values. For example:

```
{
  "provider": "custom",
  "description": "Default prices used to compute allocation between RAM and CPU. AWS pricing API data still used for total node cost.",
  "CPU": "0.031611",
  "spotCPU": "0.006655",
  "RAM": "0.004237",
  "spotRAM": "0.000892",
  "GPU": "0.95",
  "spotGPU": "",
  "storage": "0.00005479452",
  "zoneNetworkEgress": "0.01",
  "regionNetworkEgress": "0.01",
  "internetNetworkEgress": "0.143",
  "firstFiveForwardingRulesCost": "",
  "additionalForwardingRuleCost": "",
  "LBIngressDataCost": "",
  "spotLabel": "kops.k8s.io/instancegroup",
  "spotLabelValue": "spotinstance-nodes",
  "awsServiceKeyName": "AKIXXX",
  "awsSpotDataRegion": "ap-southeast-2",
  "awsSpotDataBucket": "spot-data-feed-ap-southeast-2-xxxxxxx",
  "awsSpotDataPrefix": "spotdata",
  "projectID": "xxxxxxxx",
  "athenaBucketName": "s3://x",
  "athenaRegion": "us-east-1",
  "athenaDatabase": "",
  "athenaTable": "",
  "athenaWorkgroup": "",
  "masterPayerARN": "",
  "customPricesEnabled": "",
  "defaultIdle": "",
  "azureSubscriptionID": "",
  "azureClientID": "",
  "azureClientSecret": "",
  "azureTenantID": "",
  "azureBillingRegion": "",
  "azureBillingAccount": "",
  "azureOfferDurableID": "",
  "azureStorageSubscriptionID": "",
  "azureStorageAccount": "",
  "azureStorageAccessKey": "",
  "azureStorageContainer": "",
  "azureContainerPath": "",
  "azureCloud": "",
  "currencyCode": "",
  "discount": "",
  "negotiatedDiscount": "",
  "sharedOverhead": "",
  "clusterName": "gitops-dev-1",
  "sharedNamespaces": "kube-system",
  "sharedLabelNames": "",
  "sharedLabelValues": "",
  "shareTenancyCosts": "",
  "readOnly": "",
  "editorAccess": "",
  "kubecostToken": "",
  "googleAnalyticsTag": "",
  "excludeProviderID": "",
  "defaultLBPrice": ""
}
```

In the above example, I haven't set `awsServiceKeyName` or `awsSpotDataPrefix` but they both contain values. If `awsServiceKeyName` is set then Kubecost will think you have used access keys and try to use that, hence the error `The AWS Access Key Id you provided does not exist in our records.`. Setting them to empty in values explicitly doesn't work, as they will simply not rendered by the charts and causes kubecost to use default (and useless) values.

These default values are set in opencost. https://github.com/opencost/opencost/blob/develop/configs/aws.json

To make it work, I had to log into the pod and manually delete these fields and restart the pod.

For `awsSpotDataPrefix`, if I use a data feed with no prefix it will fail to find the data since it is set to `spotdata` if not provided.

TL;DR. I think the solution is one or both of the following:
* Modify opencost to remove the default values for the fields.
* Make the helm charts to set the values to empty if they are not provided.

EDIT: Oh another thing, if you remove the values from `kubecostProductConfigs` they will not be removed from `aws.json` as it is stored in a pvc. So setting empty values might be more important.

~EDIT2: Meanwhile, setting `kubecostProductConfigs.createServiceKeySecret` to `false` seems to make IRSA work.~ Nope, need to manually remove `awsServiceKeyName` from `aws.json` for this to work.

**Comment by patrickli-foxtel** [comment]
Related to https://github.com/kubecost/cost-analyzer-helm-chart/issues/1921

**Comment by thomasvn** [comment]
Thanks for all the research done here @patrickli-foxtel! I think I agree with the proposed solution to set `"awsServiceKeyName: ""` in `/var/configs/aws.json`. I'll test the idea myself and get back to you here.

**Comment by lmouhib** [comment]
@patrickli-foxtel looks like you found the bug, this is the value I find in my deployment `","awsServiceKeyName":"AKIXXX",` which is not correct and I have not supplied.

**Comment by PriyaSivaraj1987** [comment]
I am running into same issue as well. Spot data feed doesn't accept IRSA. and there is default serviceKeyName mentioned above. Hope next version has bug fix

**Comment by thomasvn** [comment]
@patrickli-foxtel I just tested this and was only able to get Spot Data Feed + IRSA to work after setting `"awsServiceKeyName: ""` in `/var/configs/aws.json`.

@lmouhib @PriyaSivaraj1987 Could you `kubectl exec -it deploy/kubecost-cost-analyzer sh` and do the same as a workaround?

I will work to fix this issue in the OpenCost repo. Thanks again for all the help here!
